File: 543_diameter_of_binary_tree/diameterOfBinaryTree.py
class Solution:
    maxDiameter = 0
    
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def dfs(root: Optional[TreeNode]) -> int:
            if root == None:
                return 0
            left = dfs(root.left)
            right = dfs(root.right)

            self.maxDiameter = max(self.maxDiameter, left + right)
            return max(left, right) + 1

        dfs(root)

        return self.maxDiameter

File: 543_diameter_of_binary_tree/diameterOfBinaryTree.ts
function diameterOfBinaryTree(root: TreeNode | null): number {
  let max = 0;

  function getHeight(root: TreeNode | null): number {
    if (root === null) return 0;

    const left = root.left ? getHeight(root.left) : 0;
    const right = root.right ? getHeight(root.right) : 0;

    const diameter = left + right;
    max = Math.max(max, diameter);

    return Math.max(left, right) + 1;
  }

  getHeight(root);

  return max;
}
